tools/checklocks/state.go

```diff
@@ -187,6 +187,44 @@ func (l *lockState) isCompatible(other *lockState) bool {
 	return l.isSubset(other) && other.isSubset(l)
 }
 
+// join returns a new lockState of the intersection of 2 states.
+func (l *lockState) join(pc *passContext, pos token.Pos, other *lockState) *lockState {
+	if l == nil {
+		return other.fork()
+	} else if other == nil {
+		return l.fork()
+	}
+
+	rls := l.fork()
+
+	for addr, val := range rls.stored {
+		if oval, ok := other.stored[addr]; !ok || val != oval {
+			rls.modify()
+			delete(rls.stored, addr)
+		}
+	}
+
+	if !rls.isCompatible(other) {
+		pc.maybeFail(pos, "inconsistent lock states (first: %s, second: %v)", rls.String(), other.String())
```

Comment:
Reporting should probably be somebody else's job (in this case the callers). This lets you minimize the responsibilities of each function. At the moment this lets you drop pc as an argument. And not document the side effects on pc when describing join. This does require having enough info at the caller side to generate the same message. (As rls forks the caller does in this case.)